### PR TITLE
Media migration memory consumption

### PR DESCRIPTION
**Pull request by masmrlar**
With big amount of media files managed by the PIM, the media migration from version 1.3 to 1.6 is slow (and slows down during the migration process) and its quite memory consuming.

The issue is in upgrades/1.3-1.4/common/MediaMigration.php::storeLocalMedias

The foreach loop does not detach the $fileInfo entities from the entity manager.
Detaching the entities and regular flushing/clearing the entity manager helps the garbage collector to keep the memory usage low and speeds up the migration process as the entity manager does not handle the detached entities (as long as they are not reattached anymore).

| Q | A |
| --- | --- |
| Added Specs | N |
| Added Behats | N |
| Changelog updated | N |
| Review and 2 GTM | N |
| Micro Demo to the PO (Story only) | N |
| Migration script | N |
|  Tech Doc | N |

### CODE REVIEW

**Comment by fabienlem** [comment]
Hello @masmrlar,

Thank you a lot for your pull request :+1:

I copied and pasted the issue's description from the service desk support ticket in order to help our team to review the PR. We will do it soon!

Kind regards,
Fabien

File: upgrades/1.3-1.4/common/MediaMigration.php [hunk 1]
```diff
@@ -93,18 +93,52 @@ public function createFileInfoTable()
      */
     public function storeLocalMedias()
     {
+        // enable the garbage collector
+        gc_enable();
```
Comment (jjanvier) [inline upgrades/1.3-1.4/common/MediaMigration.php]: it is useful? we didn't disable it, so it's already enabled

File: upgrades/1.3-1.4/common/MediaMigration.php [hunk 2]
```diff
@@ -93,18 +93,52 @@ public function createFileInfoTable()
...
+
+        /** @var \Doctrine\ORM\EntityManagerInterface $em */
+        $em = null;
+        if (!$this->upgradeHelper->areProductsStoredInMongo()) {
```
Comment (jjanvier) [inline upgrades/1.3-1.4/common/MediaMigration.php]: fileInfo objects are always stored in MySQL
so I think you don't need this if

File: upgrades/1.3-1.4/common/MediaMigration.php [hunk 3]
```diff
@@ -93,18 +93,52 @@ public function createFileInfoTable()
...
+        $count = 0;
+        $batch = 1000;
         foreach ($finder->files()->followLinks()->in($this->mediaDirectory) as $file) {
             $fileInfo = $storer->store($file, FileStorage::CATALOG_STORAGE_ALIAS);
             $this->ormConnection->update(
                 'akeneo_file_storage_file_info',
                 ['old_file_key' => $file->getFilename()],
                 ['id'           => $fileInfo->getId()]
             );
+            // detach and unset if products are not stored in Mongo
```
Comment (jjanvier) [inline upgrades/1.3-1.4/common/MediaMigration.php]: as said before, as fileInfo objects are always in MySQL, I think you can always detach them :)

File: upgrades/1.3-1.4/common/MediaMigration.php [hunk 4]
```diff
@@ -93,18 +93,52 @@ public function createFileInfoTable()
...
+            if (null !== $em) {
+                $em->detach($fileInfo);
+            }
+            unset($fileInfo);
+            unset($file);
+
+            if (0 == $count % $batch) {
+                // flush / clear / gc
+                if (null !== $em) {
+                    $em->flush();
```
Comment (jjanvier) [inline upgrades/1.3-1.4/common/MediaMigration.php]: just to understand, why do you flush?
the file storer already flushes afaik

File: upgrades/1.3-1.4/common/MediaMigration.php [hunk 5]
```diff
@@ -93,18 +93,52 @@ public function createFileInfoTable()
...
+                    $em->clear();
+                    gc_collect_cycles();
+                }
+            }
+            $count++;
+        }
+
+        // flush / clear / gc finally
+        if (null !== $em) {
+            $em->flush();
```
Comment (jjanvier) [inline upgrades/1.3-1.4/common/MediaMigration.php]: same comment that before

File: upgrades/1.3-1.4/common/MediaMigration.php [hunk 6]
```diff
@@ -93,18 +93,52 @@ public function createFileInfoTable()
...
+            }
+            unset($fileInfo);
+            unset($file);
+
+            if (0 == $count % $batch) {
+                // flush / clear / gc
+                if (null !== $em) {
+                    $em->flush();
+                    $em->clear();
+                    gc_collect_cycles();
```
Comment (jjanvier) [inline upgrades/1.3-1.4/common/MediaMigration.php]: not sure you need to handle the gc manually
cleaning Doctrine entity manager is already a huge step :+1:

Comment (BitOne) [inline upgrades/1.3-1.4/common/MediaMigration.php]: Yeap, the GC will attempt to collect cycles by itself, so no need to call it.

File: upgrades/1.3-1.4/common/MediaMigration.php (hunk 5 above)
Comment (masmrlar) [inline upgrades/1.3-1.4/common/MediaMigration.php]: ok, going to rework this...
Thanks for the hints

**Comment by pierallard** [comment]
*info* This one is already fixed by @jmleroux in pim-community-standard https://github.com/akeneo/pim-community-standard/commit/c14e28619a578406df5f374f57c007dd063d9035

**Pull request merged**